ports/raspberrypi/common-hal/memorymap/AddressRange.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 microDev
+ * Copyright (c) 2023 Bob Abeles
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "shared-bindings/memorymap/AddressRange.h"
+
+#include "py/runtime.h"
+
+#include "hardware/regs/addressmap.h"
+
+// RP2 address map ranges, must be arranged in order by ascending start address
+addressmap_rp2_range_t rp2_ranges[] = {
+    {(uint8_t *)ROM_BASE,           0x00004000, ROM},        // boot ROM
+    {(uint8_t *)XIP_BASE,           0x00100000, XIP},        // XIP normal cache operation
+    {(uint8_t *)XIP_NOALLOC_BASE,   0x00100000, XIP},        // XIP check for hit, no update on miss
+    {(uint8_t *)XIP_NOCACHE_BASE,   0x00100000, XIP},        // XIP don't check for hit, no update on miss
+    {(uint8_t *)XIP_NOCACHE_NOALLOC_BASE, 0x00100000, XIP},  // XIP bypass cache completely
+    {(uint8_t *)XIP_CTRL_BASE,      0x00004000, IO},         // XIP control registers
+    {(uint8_t *)XIP_SRAM_BASE,      0x00004000, SRAM},       // XIP SRAM 16KB XIP cache
+    {(uint8_t *)XIP_SSI_BASE,       0x00004000, IO},         // XIP SSI registers
+    {(uint8_t *)SRAM_BASE,          0x00042000, SRAM},       // SRAM 256KB striped plus 16KB contiguous
+    {(uint8_t *)SRAM0_BASE,         0x00040000, SRAM},       // SRAM0 to SRAM3 256KB non-striped
```

Comment:
I'm a little wary of allowing SRAM access where CP is maintaining the heap. Do you *need* this? I'd prefer to be conservative to start and add areas as we have specific uses.

Reply:
Pushing back for these reasons: Overall, memorymap is a "beware, there be dragons here" sort of feature, so I prefer having a comprehensive view of the RP2 memory map and leaving its use to the discretion of the programmer. The other ports (espressif and nrf) that support memorymap do map RAM/SRAM/DRAM. Finally, it's useful to have a view into memory as a poor person's low-level debugger.